lib/hsa/mcwamp_hsa.cpp

```diff
@@ -763,11 +771,83 @@ pool_iterator::pool_iterator()
 ///
 namespace Kalmar {
 
+
+
+// Small wrapper around the hsa hardware queue (ie returned from hsa_queue_create(...).
+// This allows us to see which accelerator_view owns the hsa queue, and
+// also tracks the state of the cu mask, profiling, priority of the HW queue.
+// Rocr queues are shared by the allocated HSAQueues.  When an HSAQueue steals
+// a rocrQueue, we ensure that the hw queue has the desired cu_mask and other state.
+//
+// HSAQueue is the implementation of accelerator_view for HSA back-and.  HSAQueue
+// points to RocrQueue, or to nullptr if the HSAQueue is not currently attached to a RocrQueue.
+struct RocrQueue {
+    RocrQueue(hsa_agent_t agent, HSAQueue *hccQueue) 
+    {
+        /// Query the maximum size of the queue.
+        uint32_t queue_size = 0;
+        hsa_status_t status = hsa_agent_get_info(agent, HSA_AGENT_INFO_QUEUE_MAX_SIZE, &queue_size);
+        STATUS_CHECK(status, __LINE__);
+
+        assert (__builtin_popcount(MAX_INFLIGHT_COMMANDS_PER_QUEUE) == 1); // make sure this is power of 2.
+        assert(queue_size > MAX_INFLIGHT_COMMANDS_PER_QUEUE*2);
+
```

Comment:
Maybe we should move these asserts into HSADevice such that we would only perform them once per device rather than for every queue creation?